engines/wage/gui.cpp

```diff
@@ -0,0 +1,724 @@
+				break;
+			}
+		}
+
+		FontMan.assignFontToName(fontName, font);
+
+		debug(2, " %s", fontName.c_str());
+	}
+
+	_builtInFonts = false;
+
+	delete dat;
+}
+
+void Gui::regenCommandsMenu() {
+	_menu->regenCommandsMenu();
+}
+
+void Gui::regenWeaponsMenu() {
+	_menu->regenWeaponsMenu();
+}
+
+void Gui::processMenuShortCut(byte flags, uint16 ascii) {
+	_menu->processMenuShortCut(flags, ascii);
+}
+
+
+void Gui::mouseMove(int x, int y) {
+	if (_menu->_menuActivated) {
+		if (_menu->mouseMove(x, y))
+			_menuDirty = true;
+
+		return;
+	}
+
+	if (_inTextSelection) {
+		updateTextSelection(x, y);
+		return;
+	}
+
+	if (_consoleTextArea.contains(x, y)) {
+		if (_cursorIsArrow) {
+			CursorMan.replaceCursor(macCursorBeam, 11, 16, 3, 8, 3);
+			_cursorIsArrow = false;
+		}
+	} else if (_cursorIsArrow == false) {
+		CursorMan.replaceCursor(macCursorArrow, 11, 16, 1, 1, 3);
+		_cursorIsArrow = true;
+	}
+}
+
+Designed *Gui::mouseUp(int x, int y) {
+	if (_menu->_menuActivated) {
+		if (_menu->mouseRelease(x, y)) {
+			_sceneDirty = true;
+			_consoleDirty = true;
+			_bordersDirty = true;
+			_menuDirty = true;
+		}
+
+		return NULL;
+	}
+
+	if (_inTextSelection)
+		_inTextSelection = false;
+
+	if (_sceneArea.contains(x, y)) {
+		if (!_sceneIsActive) {
+			_sceneIsActive = true;
+			_bordersDirty = true;
+		}
+
+		for (ObjList::const_iterator it = _scene->_objs.begin(); it != _scene->_objs.end(); ++it) {
+			if ((*it)->_design->isPointOpaque(x - _sceneArea.left + kBorderWidth, y - _sceneArea.top + kBorderWidth))
+				return *it;
+		}
+
+		for (ChrList::const_iterator it = _scene->_chrs.begin(); it != _scene->_chrs.end(); ++it) {
+			if ((*it)->_design->isPointOpaque(x - _sceneArea.left + kBorderWidth, y - _sceneArea.top + kBorderWidth))
+				return *it;
+		}
+	} else if (_consoleTextArea.contains(x, y)) {
+		if (_sceneIsActive) {
+			_sceneIsActive = false;
+			_bordersDirty = true;
+		}
+	}
+
+	return NULL;
+}
+
+void Gui::mouseDown(int x, int y) {
+	if (_menu->mouseClick(x, y)) {
+		_menuDirty = true;
+	} else if (_consoleTextArea.contains(x, y)) {
+		startMarking(x, y);
+	}
+}
+
+int Gui::calcTextX(int x, int textLine) {
+	const Graphics::Font *font = getConsoleFont();
+	Common::String str = _lines[textLine];
+
+	x -= _consoleTextArea.left + kConWOverlap + kConWPadding;
+
+	for (int i = str.size(); i >= 0; i--) {
+		if (font->getStringWidth(str) < x) {
+			return i;
+		}
+
+		str.deleteLastChar();
+	}
+
+	return 0;
+}
+
+int Gui::calcTextY(int y) {
+	y -= _consoleTextArea.top + kConHOverlap + kConHPadding;
+
+	if (y < 0)
+		y = 0;
+
+	const int firstLine = _scrollPos / _consoleLineHeight;
+	int textLine = (y - _scrollPos % _consoleLineHeight) / _consoleLineHeight + firstLine;
+
+	return textLine;
+}
+
+void Gui::startMarking(int x, int y) {
+	warning("x: %d y: %d", x, y);
+	_selectionStartY = calcTextY(y);
+	_selectionStartX = calcTextX(x, _selectionStartY);
+
+	_inTextSelection = true;
+
+	warning("x: %d y: %d", _selectionStartX, _selectionStartY);
```

Comment:
I assume that the warnings here are left because this function isn't implemented completely? Please, add a TODO if this is the case